kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapter.kt

```diff
@@ -284,6 +372,333 @@ class KotlinJsonAdapterFactory : JsonAdapter.Factory {
+    if (cached != null) return cached
+
+    while (declaredFieldsIterator.hasNext()) {
+      val next = declaredFieldsIterator.next()
+      val nextSignature = next.jvmFieldSignature
+      fieldSignatureCache[nextSignature] = next
+      if (nextSignature == signatureString) {
+        return next
+      }
+    }
+
+    while (fieldIterator.hasNext()) {
+      val next = fieldIterator.next()
+      val nextSignature = next.jvmFieldSignature
+      fieldSignatureCache[nextSignature] = next
+      if (nextSignature == signatureString) {
+        return next
+      }
+    }
+
+    return null
+  }
+}
+
+internal data class ParameterData(
+    val km: KmValueParameter,
+    val index: Int,
+    val rawType: Class<*>,
+    val annotations: List<Annotation>
+) {
+  val name get() = km.name
+  val declaresDefaultValue get() = Flag.ValueParameter.DECLARES_DEFAULT_VALUE(km.flags)
+}
+
+internal data class ConstructorData(
+    val type: Class<*>,
+    val km: KmConstructor,
+    val jvm: Constructor<*>,
+    val parameters: List<ParameterData>,
+    val isDefault: Boolean
+) {
+  init {
+    jvm.isAccessible = true
+  }
+
+  fun <R> callBy(args: Map<ParameterData, Any?>): R {
+    val arguments = ArrayList<Any?>(parameters.size)
+    var mask = 0
+    val masks = ArrayList<Int>(1)
+    var index = 0
+
+    @Suppress("UseWithIndex")
+    for (parameter in parameters) {
+      if (index != 0 && index % Integer.SIZE == 0) {
+        masks.add(mask)
+        mask = 0
+      }
+
+      when {
+        args.containsKey(parameter) -> {
+          arguments.add(args[parameter])
+        }
+        parameter.declaresDefaultValue -> {
+          arguments += defaultPrimitiveValue(parameter.rawType)
+          mask = mask or (1 shl (index % Integer.SIZE))
+        }
+        else -> {
+          throw IllegalArgumentException(
+              "No argument provided for a required parameter: $parameter")
+        }
+      }
+
+      index++
+    }
+
+    if (!isDefault) {
+      @Suppress("UNCHECKED_CAST")
+      return jvm.newInstance(*arguments.toTypedArray()) as R
+    }
+
+    masks += mask
+    arguments.addAll(masks)
+
+    // DefaultConstructorMarker
+    arguments.add(null)
+
+    @Suppress("UNCHECKED_CAST")
+    return jvm.newInstance(*arguments.toTypedArray()) as R
+  }
+}
+
+internal data class PropertyData(
+    val km: KmProperty,
+    val jvmField: Field?,
+    val jvmGetter: Method?,
+    val jvmSetter: Method?,
+    val jvmAnnotationsMethod: Method?,
+    val parameter: ParameterData?
+) {
+  init {
+    jvmField?.isAccessible = true
+    jvmGetter?.isAccessible = true
+    jvmSetter?.isAccessible = true
+  }
+
+  val name get() = km.name
+
+  val javaType = jvmField?.genericType
+      ?: jvmGetter?.genericReturnType
+      ?: jvmSetter?.genericReturnType
+      ?: error("No type information available for property '${km.name}' with type '${km.returnType.canonicalName}'.")
+
+  val annotations: Set<Annotation> by lazy {
```

Comment:
a findAnnotation() method would be much more efficient

Reply:
Per above - needed due to greedy lookup of all annotations to look for json qualifiers